### PR TITLE
RDM-9262 - Support automatic Elasticsearch index creation

### PR DESCRIPTION
**Pull request by danlysiak**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-9262

### Change description ###
Support automatic Elasticsearch index creation

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/endpoint/ElasticsearchIndexController.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    @ApiOperation("Builds the Elasticsearch indices for all known case types, using their latest version. "
+        + "This API can be used to negate the need to reimport all spreadsheet definitions as part of an Elasticsearch reindex.")
+    @ApiResponses(value = {
+        @ApiResponse(code = 201, message = "Elasticsearch indices have been created successfully for all known case types"),
+        @ApiResponse(code = 400, message = "An error occurred during creation of indices"),
+        @ApiResponse(code = 404, message = "Endpoint is disabled")
+    })
+    public IndicesCreationResult createElasticsearchIndices() {
+        log.info("Creating Elasticsearch indices for latest versions of all known case types.");
+        List<CaseTypeEntity> allCaseTypes = caseTypeRepository.findAllLatestVersions();
```
Comment (mario-paniccia) [inline elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/endpoint/ElasticsearchIndexController.java:52]: a bit concerned about performance here. Would this be too heavy on the db and cause a bottleneck? Perhaps as an alternative we can get the defs one by one with a little pause between each? would need to be tested

File: elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/model/IndicesCreationResult.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toCollection;
+
+@Data
+public class IndicesCreationResult {
+
+    @JsonProperty("case_types")
+    @ApiModelProperty("Case types for which indices have been created for, grouped by jurisdiction")
+    private Map<String, Set<String>> groupedCaseTypes;
```
Comment (mario-paniccia) [inline elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/model/IndicesCreationResult.java:23]:
```suggestion
    private Map<String, Set<String>> caseTypesByJurisdiction;
```

File: elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/model/IndicesCreationResult.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    private int total;
+
+    public IndicesCreationResult(List<CaseTypeEntity> caseTypes) {
+        this.groupedCaseTypes = getGroupedCaseTypes(caseTypes);
+        this.total = caseTypes.size();
+    }
+
+    // Group the case types by jurisdiction, with both keys and values sorted
```
Comment (mario-paniccia) [inline elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/model/IndicesCreationResult.java:32]: I'd remove the comment and improve the method naming

File: elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/model/IndicesCreationResult.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    private Map<String, Set<String>> getGroupedCaseTypes(List<CaseTypeEntity> caseTypes) {
```
Comment (mario-paniccia) [inline elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/model/IndicesCreationResult.java:33]:
```suggestion
    private Map<String, Set<String>> getSortedCaseTypesGroupedByJurisdiction(List<CaseTypeEntity> caseTypes) {
```

File: charts/ccd-definition-store-api/values.yaml [hunk 5]
```diff
@@ -38,6 +38,7 @@ java:
     ELASTIC_SEARCH_DYNAMIC: true
     ELASTIC_SEARCH_CASE_INDEX_NAME_FORMAT: '%s_cases'
     ELASTIC_SEARCH_CASE_INDEX_MAPPING_FIELDS_LIMIT: 8000
+    ELASTIC_SEARCH_AUTOMATIC_INDICES_CREATION_ENABLED: true
```
Comment (mario-paniccia) [inline charts/ccd-definition-store-api/values.yaml:41]: thinking better about it, do we really need this now that this is triggered manually via an endpoint?

**Review by mario-paniccia — APPROVED** [review]
some minor comments but looks 👍

File: elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/endpoint/ElasticsearchIndexController.java (hunk 1 above)
Comment (danlysiak) [reply elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/endpoint/ElasticsearchIndexController.java:52]: As discussed, we'll make sure this gets tested in a higher env 👍

**Pull request merged**